segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/filesystem/FileSystemStorage.java

```diff
@@ -0,0 +1,377 @@
+
+    @Override
+    public CompletableFuture<SegmentProperties> create(String streamSegmentName, Duration timeout) {
+        return supplyAsync( streamSegmentName, () -> syncCreate(streamSegmentName));
+    }
+
+    @Override
+    public CompletableFuture<Void> write(SegmentHandle handle, long offset, InputStream data, int length, Duration
+            timeout) {
+        return supplyAsync(handle.getSegmentName(), () -> syncWrite(handle, offset, data, length));
+    }
+
+    @Override
+    public CompletableFuture<Void> seal(SegmentHandle handle, Duration timeout) {
+        return supplyAsync(handle.getSegmentName(), () -> syncSeal(handle));
+    }
+
+    @Override
+    public CompletableFuture<Void> concat(SegmentHandle targetHandle, long offset, String sourceSegment, Duration
+            timeout) {
+            return supplyAsync(targetHandle.getSegmentName(),
+                    () -> syncConcat(targetHandle, offset, sourceSegment));
+    }
+
+    @Override
+    public CompletableFuture<Void> delete(SegmentHandle handle, Duration timeout) {
+            return supplyAsync(handle.getSegmentName(), () -> syncDelete(handle));
+    }
+
+    //endregion
+
+    //region AutoClosable
+
+    @Override
+    public void close() {
+
+    }
+
+    //endregion
+
+    //region private sync implementation
+
+    @SneakyThrows(StreamSegmentNotExistsException.class)
+    private SegmentHandle syncOpenRead(String streamSegmentName) {
+        Path path = Paths.get(config.getFilesystemRoot(), streamSegmentName);
+
+        if (!Files.exists(path)) {
+            throw new StreamSegmentNotExistsException(streamSegmentName);
+        }
+
+        return FileSystemSegmentHandle.getReadHandle(streamSegmentName);
+    }
+
+    @SneakyThrows
+    private SegmentHandle syncOpenWrite(String streamSegmentName) {
+        Path path = Paths.get(config.getFilesystemRoot(), streamSegmentName);
+        if (!Files.exists(path)) {
+            throw new StreamSegmentNotExistsException(streamSegmentName);
+        } else if (Files.isWritable(path)) {
+            return FileSystemSegmentHandle.getWriteHandle(streamSegmentName);
+        } else {
+                return openRead(streamSegmentName).get();
+        }
+    }
+
+    @SneakyThrows(IOException.class)
+    private int syncRead(SegmentHandle handle, long offset, byte[] buffer, int bufferOffset, int length) {
+
+        Path path = Paths.get(config.getFilesystemRoot(), handle.getSegmentName());
+
+            if (Files.size(path) < offset) {
+                log.warn("Read called on segment {} at offset {}. The offset is beyond the current size of the file.",
+                        handle.getSegmentName(), offset);
+                throw new IllegalArgumentException( "Reading beyond the current size of segment");
+            }
+
+        try (FileChannel channel = FileChannel.open(path, StandardOpenOption.READ)) {
+            int bytesRead = channel.read(ByteBuffer.wrap(buffer, bufferOffset, length), offset);
```

Comment:
`Storage.read` returns and integer which means it is expected to be different than the length.